cumulusci/tasks/salesforce/update_dependencies.py

```diff
@@ -93,268 +79,165 @@ def _init_options(self, kwargs):
             return
 
-        self.logger.info("Preparing static dependencies map")
-        dependencies = self.project_config.get_static_dependencies(
-            self.options["dependencies"],
-            include_beta=self.options["include_beta"],
+        self.logger.info("Resolving dependencies...")
+        dependencies = get_static_dependencies(
+            self.dependencies,
+            self.resolution_strategy,
+            self.project_config,
             ignore_deps=self.options.get("ignore_dependencies"),
-            match_release_branch=self.options["prefer_2gp_from_release_branch"],
         )
+        self.logger.info("Collected dependencies:")
 
-        self.installed = None
-        self.uninstall_queue = []
-        self.install_queue = []
-
-        self.logger.info("Dependencies:")
-        for line in self.project_config.pretty_dependencies(dependencies):
-            self.logger.info(line)
+        for d in dependencies:
+            self.logger.info(f"    {d}")
 
-        self._process_dependencies(dependencies)
-        installs = []
-        for dep in self.install_queue:
-            if dep not in installs:
-                installs.append(dep)
+        for d in dependencies:
+            self._install_dependency(d)
 
-        self.install_queue = installs
-
-        # Reverse the uninstall queue
-        self.uninstall_queue.reverse()
-
-        self._uninstall_dependencies()
-        self._install_dependencies()
         self.org_config.reset_installed_packages()
 
-    def _process_dependencies(self, dependencies):
-        for dependency in dependencies:
-            # Process child dependencies
-            dependency_uninstalled = False
-            subdependencies = dependency.get("dependencies")
-            if subdependencies:
-                count_uninstall = len(self.uninstall_queue)
-                self._process_dependencies(subdependencies)
-                if count_uninstall != len(self.uninstall_queue):
-                    dependency_uninstalled = True
-
-            # Process namespace dependencies (managed packages)
-            if "namespace" in dependency:
-                self._process_namespace_dependency(dependency, dependency_uninstalled)
-            else:
-                # zip_url or repo dependency
-                self.install_queue.append(dependency)
-
-        if self.uninstall_queue and not self.options["allow_uninstalls"]:
-            raise TaskOptionsError(
-                "Updating dependencies would require uninstalling these packages "
-                "but uninstalls are not enabled: {}".format(
-                    ", ".join(dep["namespace"] for dep in self.uninstall_queue)
+    def _install_dependency(self, dependency):
+        if isinstance(
+            dependency, (PackageNamespaceVersionDependency, PackageVersionIdDependency)
+        ):
+            installed = False
+
+            if isinstance(dependency, PackageNamespaceVersionDependency):
```

Comment:
Is there a reason this check for the package already being installed shouldn't go in the dependency's `install` method? It seems like something we would want regardless of where in the code a package is being installed, not just for update_dependencies.

Reply:
That's an excellent idea that will make this code quite a bit cleaner, as well.

Reply:
It could even go in the core `install_package_by_*()` methods, although I feel a little leery of adding that much behavior to those primitives. I think there is a valid case to be able to reinstall a package (with different install options), although it's pretty edge-casey.

Reply:
Yeah I agree with that assessment.